app/api/access_codes.py

```diff
@@ -118,70 +141,56 @@ def query(self, view_kwargs):
             query_ = AccessCode.query.filter(AccessCode.tickets.any(id=ticket.id))
         return query_
 
-    def before_post(self, args, kwargs, data):
-        """
-        method to add user_id to view_kwargs before post
-        :param args:
-        :param kwargs:
-        :param data:
-        :return:
-        """
-        kwargs['user_id'] = current_identity.id
-
-    def before_create_object(self, data, view_kwargs):
-        if view_kwargs.get('ticket_id'):
-            ticket = safe_query(self, Ticket, 'id', view_kwargs['ticket_id'], 'ticket_id')
-            data['event_id'] = ticket.event_id
-        if view_kwargs.get('event_id'):
-            event = safe_query(self, Event, 'id', view_kwargs['event_id'], 'event_id')
-            data['event_id'] = event.id
-        elif view_kwargs.get('event_identifier'):
-            event = safe_query(self, Event, 'identifier', view_kwargs['event_identifier'], 'event_identifier')
-            data['event_id'] = event.id
-        data['user_id'] = current_identity.id
-
-    def after_create_object(self, obj, data, view_kwargs):
-        if view_kwargs.get('ticket_id'):
-            ticket = safe_query(self, Ticket, 'id', view_kwargs['ticket_id'], 'ticket_id')
-            ticket.access_codes.append(obj)
-            self.session.commit()
-
     view_kwargs = True
     decorators = (jwt_required, )
+    methods = ['GET', ]
     schema = AccessCodeSchema
     data_layer = {'session': db.session,
                   'model': AccessCode,
                   'methods': {
                       'query': query,
-                      'before_create_object': before_create_object,
-                      'after_create_object': after_create_object
                   }}
 
 
 class AccessCodeDetail(ResourceDetail):
     """
     AccessCode detail by id
     """
-    def before_get_object(self, view_kwargs):
-        if view_kwargs.get('ticket_id'):
-            ticket = safe_query(self, Ticket, 'id', view_kwargs['ticket_id'], 'ticket_id')
-            if ticket.access_code_id:
-                view_kwargs['id'] = ticket.access_code_id
-            else:
-                view_kwargs['id'] = None
+    @jwt_required
+    def before_get(self, args, kwargs):
+        """
+        Method to check if user is registered before fetching an access code
+        Only registered users can view an access code
+        :param access_code:
+        :param data:
+        :param view_kwargs:
+        :return:
+        """
+        pass
 
-    decorators = (jwt_required, )
+    def before_delete_object(self, access_code, view_kwargs):
+        """
+        Method to check for admin access before deleting an access code
+        Super-Admin/Admin can not delete an access code
+        Co-organizer can delete an access code
+        :param access_code:
+        :param view_kwargs:
+        :return:
+        """
+        if has_access('is_admin'):
+            raise ForbiddenException({'source': ''}, 'You are not authorized')
```

Comment:
Add a new function in permission manager called maybe `is_coorganizer_not_accessible_by_admin` or any other better name you can come up with. This user should check if the user is a co-organizer or organizer only and should not grant permission even if the user is admin. Then remove the `before_delete_object` method and instead use permission manager in decorators specifying the method delete.